content/blog/2023/2023-03-31-serialize-deserialize-with-jsonview.md

```diff
@@ -0,0 +1,617 @@
+    @JsonView(Views.ExternalView)
+    private String dob;
+
+    @JsonView(Views.InternalView)
+    private String loginName;
+
+    @JsonView(Views.InternalView)
+    private String loginPassword;
+    
+    private String crnNumber;
+    
+    /* More code here */
+}
+
+public class Views {
+    public static interface ExternalView {
+        
+    }
+    public static interface InternalView extends ExternalView {
+        
+    }
+}
+````
+
+As seen from the example above, Json Views help segregate confidential information from the basic ones
+by creating separate views within the same model.  
+
+### Allows control over the data exposed
+
+````java
+public class User {
+    
+    @JsonView(Views.UserSummary.class)
+    private String firstName;
+
+    @JsonView(Views.UserSummary.class)
+    private String lastName;
+
+    @JsonView(Views.UserSummary.class)
+    private String address;
+
+    @JsonView(Views.UserSummary.class)
+    private String suburb;
+
+    @JsonView(Views.UserSummary.class)
+    private String mobileNo;
+
+    @JsonView(Views.UserDetailedSummary.class)
+    private String ssnNumber;
+
+    @JsonView(Views.UserDetailedSummary.class)
+    private boolean hasBroadband;
+
+    @JsonView(Views.UserDetailedSummary.class)
+    private String broadbandConnDate;
+}
+
+public class Views { 
+    public static interface UserSummary {
+        
+    }
+    public static interface UserDetailedSummary {
+        
+    }
+}
+````
+In this example, we have created hierarchical views to have more control over the data serialized.
+Here the view `UserSummary` provides basic user details. The view `UserDetailedSummary` gives a more detailed view.
+
+### Separate views for HTTP Request Methods
+
+````java
+public class User {
+    @JsonView(Views.GetView.class)
+    private String loginName;
+
+    @JsonView(Views.GetView.class)
+    private String firstName;
+
+    @JsonView(Views.GetView.class)
+    private String lastName;
+
+    @JsonView(Views.PatchView.class)
+    private String mobileNo;
+}
+
+public class Views {
+    
+    public static interface PatchView {
+        
+    }
+
+    public static interface GetView extends PatchView {
+
+    }
+}
+````
+For the above views to apply to the HTTP methods, we will define our controller methods as below:
+
+````java
+@RestController
+public class UserController {
+
+   @PostMapping(path = "/userdetails")
+   public ResponseEntity<User> post(
+           @RequestBody @JsonView(value = Views.GetView.class) User user) {
+   return ResponseEntity.status(HttpStatus.CREATED).body(savedUser);
+}
+
+
+   @PatchMapping(path = "/userdetails/{userId}")
+   public ResponseEntity<?> patch(
+           @RequestBody @JsonView(value = Views.PatchView.class) User user) {
+   return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
+ }
+
+}
+````
+This configuration will allow only the **@JsonView mapped fields** to be updated in the POST and PATCH requests respectively. Other fields will be ignored.
```

Comment:
Could you provide a brief explanation of when to use GetView, PathView, and **both together**? I mean, if you use Views.GetView.class, will you get the field annotated with Views.PatchView.class as well?